src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs

```diff
@@ -1195,5 +1197,1805 @@ static void Main()
+        {
+            var source = @"
+using System;
+
+S s = default;
+C c = new C();
+Console.WriteLine($""{s}{c}"");
+
+ref struct S
+{
+    public static implicit operator ReadOnlySpan<char>(S s) => ""S converted"";
+}
+class C
+{
+    public static implicit operator ReadOnlySpan<char>(C s) => ""C converted"";
+}";
+
+            var interpolatedStringBuilder = GetInterpolatedStringBuilderDefinition(includeSpanOverloads: true, useDefaultParameters: false, useBoolReturns: false);
+
+            var comp = CreateCompilation(new[] { source, interpolatedStringBuilder },
+                targetFramework: TargetFramework.NetCoreApp,
+                parseOptions: TestOptions.RegularPreview);
+            var verifier = CompileAndVerify(comp, expectedOutput: @"
+value:S converted
+value:C");
+
+            verifier.VerifyIL("<top-level-statements-entry-point>", @"
+{
+  // Code size       56 (0x38)
+  .maxstack  2
+  .locals init (S V_0, //s
+                C V_1, //c
+                System.Runtime.CompilerServices.InterpolatedStringBuilder V_2)
+  IL_0000:  ldloca.s   V_0
+  IL_0002:  initobj    ""S""
+  IL_0008:  newobj     ""C..ctor()""
+  IL_000d:  stloc.1
+  IL_000e:  ldc.i4.0
+  IL_000f:  ldc.i4.2
+  IL_0010:  call       ""System.Runtime.CompilerServices.InterpolatedStringBuilder System.Runtime.CompilerServices.InterpolatedStringBuilder.Create(int, int)""
+  IL_0015:  stloc.2
+  IL_0016:  ldloca.s   V_2
+  IL_0018:  ldloc.0
+  IL_0019:  call       ""System.ReadOnlySpan<char> S.op_Implicit(S)""
+  IL_001e:  call       ""void System.Runtime.CompilerServices.InterpolatedStringBuilder.AppendFormatted(System.ReadOnlySpan<char>)""
+  IL_0023:  ldloca.s   V_2
+  IL_0025:  ldloc.1
+  IL_0026:  call       ""void System.Runtime.CompilerServices.InterpolatedStringBuilder.AppendFormatted<C>(C)""
+  IL_002b:  ldloca.s   V_2
+  IL_002d:  call       ""string System.Runtime.CompilerServices.InterpolatedStringBuilder.ToStringAndClear()""
+  IL_0032:  call       ""void System.Console.WriteLine(string)""
+  IL_0037:  ret
+}
+");
+        }
+
+        [Fact]
+        public void ExplicitUserDefinedConversionInHole()
+        {
+            var source = @"
+using System;
+
+S s = default;
+Console.WriteLine($""{s}"");
+
+ref struct S
+{
+    public static explicit operator ReadOnlySpan<char>(S s) => ""S converted"";
+}
+";
+            var interpolatedStringBuilder = GetInterpolatedStringBuilderDefinition(includeSpanOverloads: true, useDefaultParameters: false, useBoolReturns: false);
+
+            var comp = CreateCompilation(new[] { source, interpolatedStringBuilder }, targetFramework: TargetFramework.NetCoreApp, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (5,21): error CS0306: The type 'S' may not be used as a type argument
+                // Console.WriteLine($"{s}");
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "{s}").WithArguments("S").WithLocation(5, 21)
+            );
+        }
```

Comment:
>} [](http://example.com/codeflow?start=8&length=1)

Are we testing:
- scenario when AppendLiteral uses user-defined conversion
- SemanticModel APIs
- IOperation tree

Reply:
I'll add the first one. The latter 2 will come at a later time.